### PR TITLE
Make escape analysis over unexpected nodes to return conservative results

### PR DESCRIPTION
**Pull request by VSadov**
There are expressions for which ref-struct escape analysis does not make sense. Yet we keep finding ourselves in situations where we have to do it. Generally in error conditions.
So, instead of throwing "unreachable" we will return the most conservative results.

There is still an assert here so that if new bound nodes are added, there will be a chance to think whether there is a need for nontrivial handling of the node.

Fixes:#25398
Fixes:#25485

### CODE REVIEW

**Comment by VSadov** [comment]
@dotnet/roslyn-compiler - please review. Fix for a crashing bug in dev15.7.x

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs [hunk 1]
```diff
@@ -3252,5 +3252,137 @@ public static unsafe void Test(TestStruct[] ar)
...
+                Diagnostic(ErrorCode.ERR_BadSpecialByRefLocal, "var").WithArguments("S").WithLocation(12, 9),
+                // (14,9): error CS4012: Parameters or locals of type 'S' cannot be declared in async methods or lambda expressions.
+                //         var r = t.Result;
+                Diagnostic(ErrorCode.ERR_BadSpecialByRefLocal, "var").WithArguments("S").WithLocation(14, 9),
+                // (15,9): error CS8350: This combination of arguments to 'C.M(S, ref S)' is disallowed because it may expose variables referenced by parameter 't' outside of their declaration scope
+                //         M(await t, ref r);
+                Diagnostic(ErrorCode.ERR_CallArgMixing, "M(await t, ref r)").WithArguments("C.M(S, ref S)", "t").WithLocation(15, 9),
+                // (17,9): error CS1997: Since 'C.M(Task<S>)' is an async method that returns 'Task', a return keyword must not be followed by an object expression. Did you intend to return 'Task<T>'?
+                //         return default;
+                Diagnostic(ErrorCode.ERR_TaskRetNoObjectRequired, "return").WithArguments("C.M(System.Threading.Tasks.Task<S>)").WithLocation(17, 9)
```
Comment (svick) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs]: How is this error relevant here? If it's not relevant, I think the `return default;` line should be removed.

**Comment by svick** [comment]
As far as I can tell, this PR also fixes https://github.com/dotnet/roslyn/issues/25485, so I think that issue should be closed when this PR is merged.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs (hunk 1 above)
Comment (VSadov) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs]: Will remove.

**Comment by VSadov** [comment]
Yes. It fixes #25485 as well.

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs [hunk 2]
```diff
@@ -2643,20 +2655,28 @@ internal static bool CheckValEscape(SyntaxNode node, BoundExpression expr, uint
...
 
+                case BoundKind.AsOperator:
+                case BoundKind.AwaitExpression:
+                case BoundKind.ConditionalAccess:
+                case BoundKind.NullCoalescingOperator:
+                case BoundKind.ArrayAccess:
+                    // only possible in error cases (if possible at all)
+                    return false;
+
                 default:
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: > default: [](start = 15, length = 9)

Why are we moving away from throw here? The existing code is finding a lot of bugs.

Comment (VSadov) [reply src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: There is a reasonable default behavior for the nodes that we do not want explicitly handle here. The alternative is that any time we have something unexpected here we have crash in VS. 

The quality of the bugs that come from this lately is not very high - we are basically just adding cases to the switch section that does that same default behavior. 

Basically - I think the resulting "improvements" no longer justify the risk of crashes and general customer grief for this case. 
Considering that there is a safe default behavior - the assert seems more appropriate at this point.

**Comment by VSadov** [comment]
Microsoft.CodeAnalysis.Editor.UnitTests.Peek.PeekTests.TestInvokeInEmptyFile  failed. 
Could not be possibly related to this change

**Comment by VSadov** [comment]
@dotnet-bot test windows_release_unit32_prtest please

**Comment by VSadov** [comment]
@dotnet-bot test windows_release_unit32_prtest please

**Comment by VSadov** [comment]
@dotnet-bot test windows_debug_vs-integration_prtest please

**Comment by VSadov** [comment]
Thanks!!
